// fivestars-pos-sdk/packages/types/lib/features/remoteConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExperimentAssignmentRule = exports.Experiment = void 0;
var Experiment;
(function (Experiment) {
    Experiment["EARLY_CHECKIN"] = "feature_early_checkin";
    Experiment["CONSUMER_MOBILE_REWARD_REDEMPTION"] = "feature_consumer_mobile_reward_redemption";
    Experiment["EDGE_GATEWAY_LOYALTY"] = "feature_edge_gateway_loyalty";
    /**
     * mostly used to get coverage on cases where we explicitly are skipping over
     * previous experiments that no longer have an entry in the remote config
     */
    Experiment["PLACEHOLDER"] = "placeholder";
    Experiment["PLACEHOLDER_INACTIVE"] = "placeholder_inactive";
})(Experiment = exports.Experiment || (exports.Experiment = {}));
var ExperimentAssignmentRule;
(function (ExperimentAssignmentRule) {
    ExperimentAssignmentRule["LIST"] = "list";
    ExperimentAssignmentRule["RANGE"] = "range";
})(ExperimentAssignmentRule = exports.ExperimentAssignmentRule || (exports.ExperimentAssignmentRule = {}));
