// fivestars-pos-sdk/packages/fivestars-pos-sdk-pairing/src/utils/customerDeviceColor.ts
type ColorUtilNames =  'Apple' | 'Blueberry' | 'Cherry' | 'Dragonfruit' | 'Eggplant' | 'Fig' | 'Grapefruit' | 'Huckleberry' | 'Jalapeno' | 'Kale' | 
    'Lentil' | 'Mango' | 'Nectarine' | 'Okra' | 'Potato' | 'Quince' | 'Radish' | 'Strawberry' | 'Truffle' | 'Watermelon' | 'Yam' | 'Zucchini';

type ColorUtilList = Record<ColorUtilNames, string>
interface IDeviceColors {
  id: string;
  // add other properties
}
const customerDeviceColors: ColorUtilList = {
    Apple: '#79BD4B', // Light Green
    Blueberry: '#0D2C54', // Navy
    Cherry: '#B50400', // Dark Red
    Dragonfruit: '#FF2FA7', // Hot Pink
    Eggplant: '#7E149C', // Deep Purple
    Fig: '#671934', // Maroon
    Grapefruit: '#FF7A69', // Salmon
    Huckleberry: '#446DF6', // Denim
    Jalapeno: '#1FA222', // Green
    Kale: '#889180', // Army
    Lentil: '#454955', // Slate
    Mango: '#FFB400', // Mustard
    Nectarine: '#F1935A', // Peach
    Okra: '#15C13C', // Bright Green
    Potato: '#BC5D2D', // Brick Orange
    Quince: '#FCDC4D', // Yellow
    Radish: '#B74F6F', // Rosebud
    Strawberry: '#F03A47', // Red
    Truffle: '#453823', // Brown
    Watermelon: '#E14598', // Magenta
    Yam: '#DB8433', // Orange
    Zucchini: '#358F44' // Dark Green
} 

export const defaultCustomerDeviceColor = '#585858';

export const getCustomerDeviceColor = (customerDeviceDescription: keyof ColorUtilList) => {
    // dark gray as default color
    return customerDeviceColors[customerDeviceDescription] || defaultCustomerDeviceColor;
};
