// fivestars-pos-sdk/packages/fivestars-pos-sdk-messenger/src/instore-plugins/index.ts
/* istanbul ignore file */
import { 
    InstoreWindow
 } from '@fivestars-pos-sdk/types';

const instoreWindow: InstoreWindow = window as InstoreWindow;

const Plugins = instoreWindow.Plugins?.com.fivestars.instore;

/**
 * Re-export native implemenations, marking as null if the plugin was not initialized
 * This will only work properly if the plugins are initialized before the rest of the app bootstraps
 * since this is done at module load
 */
export const instorePlugins = {
    battery: Plugins?.battery?.Battery.getIsInitialized() ? Plugins.battery.Battery : null,
    bluetoothSerial: Plugins?.bluetoothserial?.BluetoothSerial.getIsInitialized()
        ? Plugins.bluetoothserial.BluetoothSerial
        : null,
    device: Plugins?.device?.Device.getIsInitialized() ? Plugins.device.Device : null,
    dgram: Plugins?.dgram?.Dgram.getIsInitialized() ? Plugins.dgram.Dgram : null,
    display: Plugins?.display?.Display.getIsInitialized() ? Plugins.display.Display : null,
    file: Plugins?.file?.File.getIsInitialized() ? Plugins.file.File : null,
    kardReader: Plugins?.kardreader?.KardReader.getIsInitialized() ? Plugins.kardreader.KardReader : null,
    ledController: Plugins?.ledcontroller?.LedController.getIsInitialized()
        ? Plugins.ledcontroller.LedController
        : null,
    metadata: Plugins?.metadata?.Metadata.getIsInitialized() ? Plugins.metadata.Metadata : null,
    navigate: Plugins?.navigate?.Navigate.getIsInitialized() ? Plugins.navigate.Navigate : null,
    networkAffinity: Plugins?.networkaffinity?.NetworkAffinity.getIsInitialized()
        ? Plugins.networkaffinity.NetworkAffinity
        : null,
    networkAlerts: Plugins?.networkalerts?.NetworkAlerts?.getIsInitialized()
        ? Plugins.networkalerts.NetworkAlerts
        : null,
    pay: Plugins?.pay?.Pay.getIsInitialized() ? Plugins.pay.Pay : null,
    prop: Plugins?.prop?.Prop.getIsInitialized() ? Plugins.prop.Prop : null,
    receiptPrinter: Plugins?.receiptprinter?.ReceiptPrinter.getIsInitialized()
        ? Plugins.receiptprinter.ReceiptPrinter
        : null,
    system: Plugins?.system?.System.getIsInitialized() ? Plugins.system.System : null,
    systemStats: Plugins?.systemstats?.SystemStats.getIsInitialized() ? Plugins.systemstats.SystemStats : null,
    thermal: Plugins?.thermal?.Thermal.getIsInitialized() ? Plugins.thermal.Thermal : null
};
