// fivestars-pos-sdk/packages/types/lib/messenger/health/health.js
"use strict";
//import { ActionType } from 'typesafe-actions';
Object.defineProperty(exports, "__esModule", { value: true });
exports.severityToLabel = exports.HealthAttributeName = exports.HealthSeverity = exports.HealthSeverityLabel = exports.HealthCardReaderStatusValue = exports.HealthStateLabel = void 0;
// import * as healthAggregationActions from './aggregation/actions';
// import { HealthAggregationState } from './aggregation/types';
// import { DiscreteHealthState } from './discrete/types';
// export type HealthAction = ActionType<typeof healthAggregationActions>;
// export interface HealthState {
//     aggregation: HealthAggregationState;
//     discrete: DiscreteHealthState;
// }
var HealthStateLabel;
(function (HealthStateLabel) {
    HealthStateLabel["BATTERY_HEALTH"] = "BatteryHealth";
    HealthStateLabel["BATTERY_LEVEL"] = "BatteryLevel";
    HealthStateLabel["BATTERY_TEMPERATURE"] = "BatteryTemperature";
    HealthStateLabel["CARD_READER_STATUS"] = "CardReaderStatus";
    HealthStateLabel["CHARGING_STATUS"] = "ChargingStatus";
    HealthStateLabel["CLOCK_SKEW"] = "ClockSkew";
    HealthStateLabel["DYNAWAVE_STATUS"] = "DynaWaveStatus";
    HealthStateLabel["EDYNAMO_STATUS"] = "EDynamoStatus";
    HealthStateLabel["PAIRING"] = "Pairing";
    HealthStateLabel["PLUGGED_IN"] = "PluggedIn";
    HealthStateLabel["WEB_SOCKET_UNACKNOWLEDGED"] = "WebSocketUnacknowledged";
    HealthStateLabel["WIFI_SIGNAL"] = "WifiSignal";
})(HealthStateLabel = exports.HealthStateLabel || (exports.HealthStateLabel = {}));
var HealthCardReaderStatusValue;
(function (HealthCardReaderStatusValue) {
    HealthCardReaderStatusValue["UNPLUGGED"] = "Unplugged";
    HealthCardReaderStatusValue["NO_USB_PERMISSION"] = "No USB Permission";
    HealthCardReaderStatusValue["REPORTING_ERROR"] = "Reporting Error";
    HealthCardReaderStatusValue["WORKING_CORRECTLY"] = "Working Correctly";
})(HealthCardReaderStatusValue = exports.HealthCardReaderStatusValue || (exports.HealthCardReaderStatusValue = {}));
var HealthSeverityLabel;
(function (HealthSeverityLabel) {
    HealthSeverityLabel["NORMAL"] = "normal";
    HealthSeverityLabel["MINOR"] = "minor";
    HealthSeverityLabel["MAJOR"] = "major";
    HealthSeverityLabel["CRITICAL"] = "critical";
})(HealthSeverityLabel = exports.HealthSeverityLabel || (exports.HealthSeverityLabel = {}));
var HealthSeverity;
(function (HealthSeverity) {
    HealthSeverity[HealthSeverity["NORMAL"] = 0] = "NORMAL";
    HealthSeverity[HealthSeverity["MINOR"] = 1] = "MINOR";
    HealthSeverity[HealthSeverity["MAJOR"] = 2] = "MAJOR";
    HealthSeverity[HealthSeverity["CRITICAL"] = 3] = "CRITICAL";
})(HealthSeverity = exports.HealthSeverity || (exports.HealthSeverity = {}));
var HealthAttributeName;
(function (HealthAttributeName) {
    HealthAttributeName["CARD_READER_HARDWARE_NAME"] = "card_reader_hardware_name";
})(HealthAttributeName = exports.HealthAttributeName || (exports.HealthAttributeName = {}));
var severityToLabel = function (sev) {
    switch (sev) {
        case HealthSeverity.NORMAL:
            return HealthSeverityLabel.NORMAL;
        case HealthSeverity.MINOR:
            return HealthSeverityLabel.MINOR;
        case HealthSeverity.MAJOR:
            return HealthSeverityLabel.MAJOR;
        case HealthSeverity.CRITICAL:
            return HealthSeverityLabel.CRITICAL;
    }
};
exports.severityToLabel = severityToLabel;
