// fivestars-pos-sdk/packages/fivestars-pos-sdk-auth/lib/src/loyalty/types.js
"use strict";
// import { ApiBusiness } from '@instore/redux/business/types';
// import { IApiFeatureSetting } from '@instore/redux/features/types';
// import { HardwareVersion, Peripheral } from '@instore/redux/system/types';
// import { ClientType, Product } from '@instore/utils/enums';
// export interface LoginLoyaltyPayload {
//     app_version: string; // [INSTORE-1510] deprecated
//     client_type: ClientType;
//     description: string | null;
//     device_type: string;
//     email: string;
//     hardware_identifier: string;
//     password: string;
//     platform: string;
//     product: Product;
//     software_version: string; // [INSTORE-1510] deprecated
//     version: string;
// }
// export interface LoginLoyaltyResponse {
//     apikey: string;
//     apisecret: string;
//     description: string;
// }
// export interface GetBusinessDetailsResponse {
//     featureFlags: IApiFeatureSetting[];
//     meta: {
//         embed: {
//             business: ApiBusiness;
//         };
//     };
// }
// // There is definitely more to this, but device description is all we use it for
// export interface HeartbeatReponse {
//     description: string;
// }
// export interface EdgeAuthResponse {
//     access_token: string;
//     access_expires_at: number;
//     refresh_token: string;
// }
// export interface JwtResponse {
//     token: string;
//     type: 'jwt';
// }
// export interface FsJwt {
//     // standard jwt claims
//     aud: string;
//     iss: string;
//     exp: number;
//     iat: number;
//     sub: string;
//     // fs added
//     ['fs:b_sid']: string;
//     ['fs:b_uid']: string;
//     ['fs:ut']: string;
//     ['fs:v']: number;
// }
// export interface EdgeAuthDetails {
//     decodedJwt: FsJwt;
//     rawJwt: string;
//     refreshToken: string;
// }
// export interface ShippedHardwareResponse {
//     software_type: string;
//     hardware_version: HardwareVersion;
//     peripherals: Peripheral[] | undefined;
// }
// export interface SaleResponse {
//     state: string;
//     uid: string;
// }
// export interface SendAppDownloadLinkPayload {
//     account_uid: string;
// }
