// fivestars-pos-sdk/packages/fivestars-pos-sdk-auth/lib/src/loyalty/urls.js
"use strict";
// import { SagaIterator } from 'redux-saga';
// import { select } from 'redux-saga/effects';
// import { CONFIG } from '@app/config';
// import { edgeGatewayLoyaltyEnabled } from '@instore/redux/features/selectors';
// import { autoBind } from '@instore/utils/autoBind';
// import { UrlProviderType } from '../makeApiRequest/types';
// import { UrlProvider } from '../types';
// export const URL = {
//     CUSTOMER: 'customers',
//     EDGE_LOGIN: 'edge/auth/login',
//     EDGE_REFRESH: 'edge/auth/refresh',
//     PRODUCT: CONFIG.LOYALTY_API.PRODUCT_PATH,
//     S3_LOG_CREDENTIALS: 's3-log-credentials'
// };
// class LoyaltyUrlsImpl {
//     private static BASE_URL = `${CONFIG.LOYALTY_API.SCHEME}://${CONFIG.LOYALTY_API.HOST}:${CONFIG.LOYALTY_API.PORT}`;
//     private static EDGE_BASE_URL = `${CONFIG.LOYALTY_API.SCHEME}://${CONFIG.EDGE_API.HOST}`;
//     private static BASE_PATH = `${CONFIG.LOYALTY_API.PATH}`;
//     private static BUSINESS_BASE_PATH = `${LoyaltyUrlsImpl.BASE_PATH}/businesses`;
//     private baseBusinessPath = '';
//     private apiKey = '';
//     constructor() {
//         autoBind(this);
//     }
//     public setSoftwareId(softwareId: string) {
//         this.setCredentials(softwareId, '');
//     }
//     public setCredentials(softwareId: string, apiKey: string) {
//         this.baseBusinessPath = `${LoyaltyUrlsImpl.BUSINESS_BASE_PATH}/${softwareId}`;
//         this.apiKey = apiKey;
//     }
//     public getApiKey(): string {
//         return this.apiKey;
//     }
//     public getBusinessPath(additional?: string): string {
//         if (additional) {
//             return `${this.baseBusinessPath}/${additional}`;
//         }
//         return this.baseBusinessPath;
//     }
//     public getRootApiPath(additional?: string): string {
//         if (additional) {
//             return `${LoyaltyUrlsImpl.BASE_PATH}/${additional}`;
//         }
//         return `${LoyaltyUrlsImpl.BASE_PATH}`;
//     }
//     public getLoyaltyUrlBase(): string {
//         return LoyaltyUrlsImpl.BASE_URL;
//     }
//     public getEdgeUrlBase(): string {
//         return LoyaltyUrlsImpl.EDGE_BASE_URL;
//     }
// }
// export const LoyaltyUrls = new LoyaltyUrlsImpl();
// export class LoyaltyUrlProvider implements UrlProvider {
//     constructor() {
//         autoBind(this);
//     }
//     *provide(): SagaIterator<string> {
//         const useEdge: boolean = yield select(edgeGatewayLoyaltyEnabled);
//         if (useEdge) {
//             return LoyaltyUrls.getEdgeUrlBase();
//         }
//         return LoyaltyUrls.getLoyaltyUrlBase();
//     }
//     getType(): UrlProviderType {
//         return UrlProviderType.LOYALTY;
//     }
// }
// export class AlwaysLoyaltyUrlProvider implements UrlProvider {
//     constructor() {
//         autoBind(this);
//     }
//     *provide(): SagaIterator<string> {
//         return LoyaltyUrls.getLoyaltyUrlBase();
//     }
//     getType(): UrlProviderType {
//         return UrlProviderType.LOYALTY_ONLY;
//     }
// }
