// fivestars-pos-sdk/packages/types/lib/system.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoginStatus = void 0;
// import { ActionType } from 'typesafe-actions';
// import { Product, ShutdownSource } from '@instore/utils/enums';
// import { InternetStatus } from '@instore/utils/internetStatus';
// import * as systemActions from './actions';
// export enum SystemJob {
//     NONE = 'none',
//     RESTART = 'restart',
//     INSTALL = 'install',
//     DOWNLOAD = 'download'
// }
var LoginStatus;
(function (LoginStatus) {
    LoginStatus["FAILED_BAD_CREDENTIALS"] = "FAILED_BAD_CREDENTIALS";
    LoginStatus["FAILED_NETWORK_ERROR"] = "FAILED_NETWORK_ERROR";
    LoginStatus["PENDING"] = "PENDING";
    LoginStatus["NONE"] = "";
    LoginStatus["SUCCEEDED"] = "SUCCEEDED";
})(LoginStatus = exports.LoginStatus || (exports.LoginStatus = {}));
// export enum HardwareVersion {
//     ONE = '1.0',
//     ONE_FIVE = '1.5',
//     TWO = '2.0'
// }
// export enum Peripheral {
//     ETHERNET = 'ethernet',
//     LTE = 'lte',
//     NFC = 'nfc'
// }
// export type ShippedHardwareConfiguration = {
//     product: Product | null;
//     hardwareVersion: HardwareVersion | null;
//     peripherals: Peripheral[];
// };
// export interface SystemState {
//     clockSkewMs: number;
//     internetStatus: InternetStatus;
//     job: SystemJob;
//     lastVGSReloadAttempt: moment.Moment | null;
//     logUploadInProgress: boolean;
//     nightlyRestartReady: boolean;
//     pairedDeviceInternetStatus: InternetStatus;
//     pairedDeviceMessengerVersion: string;
//     product: Product; // separate from shipped hardware config, set to CONFIG.PRODUCT
//     shippedHardwareConfiguration: ShippedHardwareConfiguration;
//     shutdownSource: ShutdownSource;
//     // flag to enable writing to indexeddb
//     indexedDbEnabled: boolean;
// }
// export type SystemAction = ActionType<typeof systemActions>;
