// fivestars-pos-sdk/packages/fivestars-pos-sdk-transaction/src/transactions.ts
import { TransactionsModel } from '@fivestars-pos-sdk/fivestars-pos-sdk-storage'
import { InitiatePaymentOptions, FailedOcrPayload, TransactionType } from './utils'

export const LOG_PREFIX = 'Transactions'

export class Transactions {

    constructor() {
        console.info(`${LOG_PREFIX}::constructor`);
    }

    public initiatePayment(options: InitiatePaymentOptions): void {
        TransactionsModel.transactionType = TransactionType.PAYMENT
    }

    public initiateRefund(options: InitiatePaymentOptions): void {
        TransactionsModel.transactionType = TransactionType.REFUND
    }

    public isTransactionInProgress(): boolean {
        const isTransactionInProgress: boolean =
            TransactionsModel.initialTransactionAmount.getAmount() !== 0 ||
            TransactionsModel.refundAmount.getAmount() !== 0 ||
            TransactionsModel.isEarlyCheckin

        return isTransactionInProgress;
    }

    public isTransactionCompleted = (): boolean => {
        return TransactionsModel.transactionCompletedTime !== null
    }    

    public isEarlyCheckin(): boolean {
        return TransactionsModel.isEarlyCheckin
    }

    public failedOcrPayload(failedOcrPayload: FailedOcrPayload ): void {
        TransactionsModel.failedOcrPayload = failedOcrPayload
    }
}
