// fivestars-pos-sdk/packages/types/src/system.ts
import * as moment from 'moment';
// import { ActionType } from 'typesafe-actions';

// import { Product, ShutdownSource } from '@instore/utils/enums';
// import { InternetStatus } from '@instore/utils/internetStatus';

// import * as systemActions from './actions';

// export enum SystemJob {
//     NONE = 'none',
//     RESTART = 'restart',
//     INSTALL = 'install',
//     DOWNLOAD = 'download'
// }

export enum LoginStatus {
    FAILED_BAD_CREDENTIALS = 'FAILED_BAD_CREDENTIALS',
    FAILED_NETWORK_ERROR = 'FAILED_NETWORK_ERROR',
    PENDING = 'PENDING',
    NONE = '',
    SUCCEEDED = 'SUCCEEDED'
}

// export enum HardwareVersion {
//     ONE = '1.0',
//     ONE_FIVE = '1.5',
//     TWO = '2.0'
// }

// export enum Peripheral {
//     ETHERNET = 'ethernet',
//     LTE = 'lte',
//     NFC = 'nfc'
// }

// export type ShippedHardwareConfiguration = {
//     product: Product | null;
//     hardwareVersion: HardwareVersion | null;
//     peripherals: Peripheral[];
// };

// export interface SystemState {
//     clockSkewMs: number;
//     internetStatus: InternetStatus;
//     job: SystemJob;
//     lastVGSReloadAttempt: moment.Moment | null;
//     logUploadInProgress: boolean;
//     nightlyRestartReady: boolean;
//     pairedDeviceInternetStatus: InternetStatus;
//     pairedDeviceMessengerVersion: string;
//     product: Product; // separate from shipped hardware config, set to CONFIG.PRODUCT
//     shippedHardwareConfiguration: ShippedHardwareConfiguration;
//     shutdownSource: ShutdownSource;

//     // flag to enable writing to indexeddb
//     indexedDbEnabled: boolean;
// }

// export type SystemAction = ActionType<typeof systemActions>;
