// fivestars-pos-sdk/packages/fivestars-pos-sdk-auth/src/auth.ts

import axios from 'axios';
import { Core } from "@fivestars-pos-sdk/fivestars-pos-sdk-core"
import { AuthModel, PairingModel } from '@fivestars-pos-sdk/fivestars-pos-sdk-storage'
import { ILogin } from '@fivestars-pos-sdk/types'

export const LOG_PREFIX = 'Auth';

export class Auth {
    private core: Core

    constructor() {
        this.core = new Core()
        // this.core.on('WEB_SOCKET_AUTHENTICATION_SUCCESSFUL', (event: WebSocket.MessageEvent) => {
        //     console.log('in pairing::WEB_SOCKET_AUTHENTICATION_SUCCESSFUL')
        // })
        console.info(`${LOG_PREFIX}::constructor`);
    }

    public async login(info: ILogin): Promise<boolean> {
       try {                
            const API_TIMEOUT = 43000   

            const loyalty = await axios.post('https://loyalty-api.nerfstars.com/api/unified/businesses/PAYINSTORE20/merchant-tablets', {
                app_version: "2.5.0",
                client_type: "browser",
                description: null,
                device_type: "web",
                email: info.email,
                hardware_identifier: "9026",
                password: info.password,
                platform: "",
                product: "mreg",
                software_version: "2.5.0",
                version: "2.5.0"
            })

            AuthModel.loyaltyCredentials = {
                apiKey: loyalty.data.apikey,
                apiSecret: loyalty.data.apisecret,
                softwareId: info.softwareId
            }

            PairingModel.deviceDescription = loyalty.data.description

            console.info('device description', PairingModel.deviceDescription)
            console.info('loyalty credentials', AuthModel.loyaltyCredentials)
            console.info(`${LOG_PREFIX}::login sucessful login`)        

            return Promise.resolve(true)
        } 
        catch (error) {
            console.info(`${LOG_PREFIX}::login `, error)
            return Promise.resolve(false)
        }
    }
}
