// fivestars-pos-sdk/packages/fivestars-pos-sdk-messenger/src/messenger/actions/paired/cpay.ts
// import { AnyAction } from 'redux';
// import { createAction } from 'typesafe-actions';

// import { CONFIG } from '@app/config';
// import { CardReaderStatus } from '@instore/hardware/cardReader/types';
// import { PreviousTransactionState } from '@instore/redux/previousTransaction/types';
// import { CardCompliantDetails, CardError, IApiTransaction } from '@instore/redux/transaction/types';
// import { CardInputType, PayStatus } from '@instore/utils/enums';
// import { IMoney } from '@instore/utils/money';

// import { sendPaired } from './shared';

// export const sendCardError = (cardError: CardError | null, lastCardInputType: CardInputType) =>
//     sendPaired({
//         action: 'pay_card_error',
//         confirmation: null,
//         data: { cardError, lastCardInputType },
//         skipAcknowledgment: false
//     });

// export const sendTipAmount = (tipAmount: IMoney) =>
//     sendPaired({
//         action: 'pay_tip',
//         confirmation: null,
//         data: { value: tipAmount.getAmount() },
//         skipAcknowledgment: false
//     });

// export const sendPayStatus = (status: PayStatus, confirmation: AnyAction | null = null) =>
//     sendPaired({
//         action: 'pay_status',
//         confirmation,
//         data: { status },
//         skipAcknowledgment: false
//     });

// export const sendCardCompliantDetails = (cardCompliantDetails: CardCompliantDetails) =>
//     sendPaired({
//         action: 'pay_card',
//         confirmation: null,
//         data: {
//             accountUid: cardCompliantDetails.accountUid,
//             cardBrand: cardCompliantDetails.cardBrand,
//             cardUid: cardCompliantDetails.cardUid,
//             createdAt: cardCompliantDetails.createdAt,
//             fullName: cardCompliantDetails.fullName,
//             last4: cardCompliantDetails.last4,
//             token: cardCompliantDetails.tokenUid
//         },
//         skipAcknowledgment: false
//     });

// export const sendPayTransaction = (transaction: any) =>
//     sendPaired({ action: 'pay_transaction', confirmation: null, data: { transaction }, skipAcknowledgment: false });

// export const sentPayCustomer = createAction('MESSENGER_SENT_PAY_CUSTOMER');

// export const sendPayCustomer = (customer: any) =>
//     sendPaired({
//         action: 'pay_customer',
//         confirmation: sentPayCustomer(),
//         data: { customer },
//         skipAcknowledgment: false
//     });

// export const sendPayRewards = (rewards: any) =>
//     sendPaired({ action: 'pay_rewards', confirmation: null, data: rewards, skipAcknowledgment: false });

// export const sendBatteryStatus = (batteryStatus: any) =>
//     sendPaired({
//         action: 'battery_state_update',
//         confirmation: null,
//         data: { ...batteryStatus, sourceProduct: CONFIG.PRODUCT },
//         skipAcknowledgment: true
//     });

// export const sendPotentialDuplicateTransaction = (previousTransaction: PreviousTransactionState) =>
//     sendPaired({
//         action: 'potential_duplicate_transaction',
//         confirmation: null,
//         data: {
//             previousTransaction: {
//                 ...previousTransaction,
//                 transaction: {
//                     ...previousTransaction.transaction,
//                     cashAmount: previousTransaction.transaction.cashAmount.getAmount(),
//                     initialTransactionAmount: previousTransaction.transaction.initialTransactionAmount.getAmount(),
//                     // prettier-ignore
//                     initialTransactionAmountMinusReward:
//                             previousTransaction.transaction.initialTransactionAmountMinusReward.getAmount(),
//                     refundAmount: previousTransaction.transaction.refundAmount.getAmount(),
//                     tip: previousTransaction.transaction.tip.getAmount()
//                 }
//             }
//         },
//         skipAcknowledgment: false
//     });

// export const sendRefundableTransactions = (refundableTransactions: IApiTransaction[]) =>
//     sendPaired({
//         action: 'refundable_transactions',
//         confirmation: null,
//         data: { refundableTransactions },
//         skipAcknowledgment: false
//     });

// export const selectedRefundableTransaction = (refundableTransaction: IApiTransaction) =>
//     sendPaired({
//         action: 'selected_refundable_transaction',
//         confirmation: null,
//         data: { refundableTransaction },
//         skipAcknowledgment: false
//     });

// // send mPay the transaction that cPay has refunded so mPay has all the information
// // needed to refund the appropriate number of points
// export const sendRefundedTransaction = (refundedTransaction: IApiTransaction) =>
//     sendPaired({
//         action: 'refunded_transaction',
//         confirmation: null,
//         data: { refundedTransaction },
//         skipAcknowledgment: false
//     });

// export const sendCardRemoved = () =>
//     sendPaired({ action: 'pay_card_removed', confirmation: null, data: {}, skipAcknowledgment: false });

// export const sendCardReaderStatus = (cardReaderStatus: CardReaderStatus) =>
//     sendPaired({
//         action: 'full_card_reader_status',
//         confirmation: null,
//         data: { cardReaderStatus },
//         skipAcknowledgment: true
//     });

// export const sendTransactionStartResponse = (sessionId: string) =>
//     sendPaired({
//         action: 'transaction_start_response',
//         confirmation: null,
//         data: {
//             sessionId
//         },
//         skipAcknowledgment: false
//     });

// export const sendTransactionCardInsertedResponse = () =>
//     sendPaired({
//         action: 'transaction_start_card_inserted_response',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendTransactionNotOnSplashResponse = () =>
//     sendPaired({
//         action: 'transaction_not_on_splash_response',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendEarlyCheckinStart = () =>
//     sendPaired({
//         action: 'early_checkin_start',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendEarlyCheckinCancel = () =>
//     sendPaired({
//         action: 'early_checkin_cancel',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendEarlyCheckinCustomerDone = () =>
//     sendPaired({
//         action: 'early_checkin_customer_done',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendEarlyCheckinSignOut = () =>
//     sendPaired({
//         action: 'early_checkin_sign_out',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendApproveTotal = () =>
//     sendPaired({
//         action: 'approve_total',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });

// export const sendDeclineTotal = () =>
//     sendPaired({
//         action: 'decline_total',
//         confirmation: null,
//         data: {},
//         skipAcknowledgment: false
//     });
