// fivestars-pos-sdk/packages/fivestars-pos-sdk-transaction/lib/transactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transactions = exports.LOG_PREFIX = void 0;
var fivestars_pos_sdk_storage_1 = require("@fivestars-pos-sdk/fivestars-pos-sdk-storage");
var utils_1 = require("./utils");
exports.LOG_PREFIX = 'Transactions';
var Transactions = /** @class */ (function () {
    function Transactions() {
        this.isTransactionCompleted = function () {
            return fivestars_pos_sdk_storage_1.TransactionsModel.transactionCompletedTime !== null;
        };
        console.info(exports.LOG_PREFIX + "::constructor");
    }
    Transactions.prototype.initiatePayment = function (options) {
        fivestars_pos_sdk_storage_1.TransactionsModel.transactionType = utils_1.TransactionType.PAYMENT;
    };
    Transactions.prototype.initiateRefund = function (options) {
        fivestars_pos_sdk_storage_1.TransactionsModel.transactionType = utils_1.TransactionType.REFUND;
    };
    Transactions.prototype.isTransactionInProgress = function () {
        var isTransactionInProgress = fivestars_pos_sdk_storage_1.TransactionsModel.initialTransactionAmount.getAmount() !== 0 ||
            fivestars_pos_sdk_storage_1.TransactionsModel.refundAmount.getAmount() !== 0 ||
            fivestars_pos_sdk_storage_1.TransactionsModel.isEarlyCheckin;
        return isTransactionInProgress;
    };
    Transactions.prototype.isEarlyCheckin = function () {
        return fivestars_pos_sdk_storage_1.TransactionsModel.isEarlyCheckin;
    };
    Transactions.prototype.failedOcrPayload = function (failedOcrPayload) {
        fivestars_pos_sdk_storage_1.TransactionsModel.failedOcrPayload = failedOcrPayload;
    };
    return Transactions;
}());
exports.Transactions = Transactions;
