// fivestars-pos-sdk/packages/fivestars-pos-sdk-messenger/lib/src/messenger/actions/transports/webSocket.js
"use strict";
// import { createAction } from 'typesafe-actions';
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscriberAuthenticated = exports.pongTimeout = exports.pongReceived = exports.authenticationSuccessful = exports.authenticationFailure = exports.closeEvent = exports.connectEvent = exports.failedToReconnect = exports.reconnectRequest = exports.errorEvent = void 0;
// export const errorEvent = createAction('WEB_SOCKET_ERROR_EVENT', resolve => (message: string) => resolve(message));
var errorEvent = function (message) { return ({ action: 'WEB_SOCKET_ERROR_EVENT', data: message }); };
exports.errorEvent = errorEvent;
// export const reconnectRequest = createAction(
//     'WEB_SOCKET_RECONNECT_REQUEST',
//     resolve => (reason: WebSocketReconnectReason) => resolve({ reason })
// );
var reconnectRequest = function (reason) { return ({ action: 'WEB_SOCKET_RECONNECT_REQUEST', data: reason }); };
exports.reconnectRequest = reconnectRequest;
// export const failedToReconnect = createAction('WEB_SOCKET_FAILED_TO_RECONNECT');
var failedToReconnect = function () { return ({ action: 'WEB_SOCKET_FAILED_TO_RECONNECT' }); };
exports.failedToReconnect = failedToReconnect;
// export const connectEvent = createAction('WEB_SOCKET_CONNECT_EVENT');
var connectEvent = function () { return ({ action: 'WEB_SOCKET_CONNECT_EVENT' }); };
exports.connectEvent = connectEvent;
// export const closeEvent = createAction('WEB_SOCKET_CLOSE_EVENT', resolve => (reason: string) => resolve(reason));
var closeEvent = function (reason) { return ({ action: 'WEB_SOCKET_CLOSE_EVENT', data: reason }); };
exports.closeEvent = closeEvent;
// export const authenticationFailure = createAction('WEB_SOCKET_AUTHENTICATION_FAILURE');
var authenticationFailure = function () { return ({ action: 'WEB_SOCKET_AUTHENTICATION_FAILURE' }); };
exports.authenticationFailure = authenticationFailure;
// export const authenticationSuccessful = createAction('WEB_SOCKET_AUTHENTICATION_SUCCESSFUL');
var authenticationSuccessful = function () { return ({ action: 'WEB_SOCKET_AUTHENTICATION_SUCCESSFUL' }); };
exports.authenticationSuccessful = authenticationSuccessful;
// export const pongReceived = createAction('WEB_SOCKET_PONG_RECEIVED', resolve => () => resolve(LOG_FILTER));
var pongReceived = function () { return ({ action: 'WEB_SOCKET_PONG_RECEIVED' }); };
exports.pongReceived = pongReceived;
// export const pongTimeout = createAction('WEB_SOCKET_PONG_TIMEOUT');
var pongTimeout = function () { return ({ action: 'WEB_SOCKET_PONG_TIMEOUT' }); };
exports.pongTimeout = pongTimeout;
// export const subscriberAuthenticated = createAction('WEB_SOCKET_SUBSCRIBER_AUTHENTICATED');
var subscriberAuthenticated = function () { return ({ action: 'WEB_SOCKET_SUBSCRIBER_AUTHENTICATED' }); };
exports.subscriberAuthenticated = subscriberAuthenticated;
