// fivestars-pos-sdk/packages/fivestars-pos-sdk-storage/lib/models/transactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionsModel = void 0;
var types_1 = require("@fivestars-pos-sdk/types");
var fivestars_pos_sdk_utils_1 = require("@fivestars-pos-sdk/fivestars-pos-sdk-utils");
exports.TransactionsModel = {
    approveTotalNeeded: false,
    cardAuthorizationAttempts: 0,
    cardCompliantDetails: null,
    cardError: null,
    cardImpression: null,
    cashAmount: new fivestars_pos_sdk_utils_1.Money(0),
    cashDiscount: new fivestars_pos_sdk_utils_1.Money(0),
    cdAdjustment: new fivestars_pos_sdk_utils_1.Money(0),
    earlyCheckinModal: types_1.EarlyCheckinModal.NONE,
    failedOcr: null,
    failedOcrPayload: null,
    groupRefundFailures: [],
    groupRefundSuccesses: [],
    groupUid: null,
    groupUidStatus: types_1.Status.NONE,
    hostSplitPayment: false,
    idempotencyKey: null,
    initialTransactionAmount: new fivestars_pos_sdk_utils_1.Money(0),
    initialTransactionAmountMinusReward: new fivestars_pos_sdk_utils_1.Money(0),
    initiateTransactionSent: false,
    isConsumerMobileAppDownloadTextSent: false,
    isEarlyCheckin: false,
    options: {
        cashEnabled: true,
        groupUid: null,
        manualTransaction: false,
        showRewards: true,
        splitPayment: false,
        splitPaymentNumber: 1
    },
    orderSummaryOrderTabSelected: true,
    paymentType: '',
    pendingPayment: null,
    pendingRefund: null,
    points: 0,
    receiptUid: '',
    refundAmount: new fivestars_pos_sdk_utils_1.Money(0),
    refundStatus: types_1.Status.NONE,
    refundableTransactions: [],
    rewardApplied: false,
    selectedRefund: null,
    selectedReward: {
        active: false,
        itemName: '',
        perkShortName: '',
        pointCost: 0,
        type: '',
        uid: '',
        value: 0
    },
    sessionId: '',
    showCancelSplitPaymentModal: false,
    splitPayment: false,
    splitPaymentAmount: new fivestars_pos_sdk_utils_1.Money(0),
    splitPaymentCashDiscount: new fivestars_pos_sdk_utils_1.Money(0),
    splitPaymentCdAdjustment: new fivestars_pos_sdk_utils_1.Money(0),
    splitPaymentEqualSplitTotal: new fivestars_pos_sdk_utils_1.Money(0),
    splitPaymentEqualSplits: 0,
    splitPaymentNumber: 1,
    splitPaymentStates: [],
    splitPaymentTotalSplits: 0,
    tip: new fivestars_pos_sdk_utils_1.Money(0),
    transactionCompletedTime: null,
    transactionCreatedTime: null,
    transactionSent: false,
    transactionTotalUid: null,
    transactionType: types_1.TransactionType.NONE,
    transactionUid: '',
    transactionsCount: 0,
    transactionsList: [],
    transactionsListStatus: types_1.Status.NONE
};
