// fivestars-pos-sdk/packages/types/lib/instore-plugins/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReceiptPrinterModule = exports.KardReaderModule = exports.ORIENTATION = void 0;
var ORIENTATION;
(function (ORIENTATION) {
    ORIENTATION["PORTRAIT"] = "portrait";
    ORIENTATION["LANDSCAPE"] = "landscape";
})(ORIENTATION = exports.ORIENTATION || (exports.ORIENTATION = {}));
var KardReaderModule;
(function (KardReaderModule) {
    var PayloadType;
    (function (PayloadType) {
        PayloadType["MAGTEK"] = "Magtek";
        PayloadType["PAX"] = "Pax";
    })(PayloadType = KardReaderModule.PayloadType || (KardReaderModule.PayloadType = {}));
    /**
     * Topics used to subscribe to card reader status updates from the edynamo plugin
     */
    var CardReaderStatusTopic;
    (function (CardReaderStatusTopic) {
        CardReaderStatusTopic["BATTERY_LEVEL"] = "BATTERY_LEVEL";
        CardReaderStatusTopic["CONNECTED"] = "CONNECTED";
        CardReaderStatusTopic["CARD_INSERTED"] = "CARD_INSERTED";
        CardReaderStatusTopic["USB_PERMISSION"] = "USB_PERMISSION";
    })(CardReaderStatusTopic = KardReaderModule.CardReaderStatusTopic || (KardReaderModule.CardReaderStatusTopic = {}));
    var CardReaderTransactionStatusEvent;
    (function (CardReaderTransactionStatusEvent) {
        CardReaderTransactionStatusEvent["NO_EVENTS"] = "NO_EVENTS";
        CardReaderTransactionStatusEvent["CARD_INSERTED"] = "CARD_INSERTED";
        CardReaderTransactionStatusEvent["CARD_ERROR"] = "CARD_ERROR";
        CardReaderTransactionStatusEvent["TRANSACTION_PROGRESS_CHANGE"] = "TRANSACTION_PROGRESS_CHANGE";
        CardReaderTransactionStatusEvent["WAITING_FOR_USER_RESPONSE"] = "WAITING_FOR_USER_RESPONSE";
        CardReaderTransactionStatusEvent["TIMEOUT"] = "TIMEOUT";
        CardReaderTransactionStatusEvent["TRANSACTION_TERMINATED"] = "TRANSACTION_TERMINATED";
        CardReaderTransactionStatusEvent["HOST_CANCELLED_TRANSACTION"] = "HOST_CANCELLED_TRANSACTION";
        CardReaderTransactionStatusEvent["CARD_REMOVED"] = "CARD_REMOVED";
    })(CardReaderTransactionStatusEvent = KardReaderModule.CardReaderTransactionStatusEvent || (KardReaderModule.CardReaderTransactionStatusEvent = {}));
    var CardReaderTransactionStatusProgress;
    (function (CardReaderTransactionStatusProgress) {
        CardReaderTransactionStatusProgress["NO_TRANSACTION_IN_PROGRESS"] = "NO_TRANSACTION_IN_PROGRESS";
        CardReaderTransactionStatusProgress["WAITING_FOR_USER_TO_INSERT_CARD"] = "WAITING_FOR_USER_TO_INSERT_CARD";
        CardReaderTransactionStatusProgress["POWERING_UP_CARD"] = "POWERING_UP_CARD";
        CardReaderTransactionStatusProgress["SELECTING_APPLICATION"] = "SELECTING_APPLICATION";
        CardReaderTransactionStatusProgress["WAITING_USER_LANGUAGE_SELECTION"] = "WAITING_USER_LANGUAGE_SELECTION";
        CardReaderTransactionStatusProgress["WAITING_USER_APP_SELECTION"] = "WAITING_USER_APP_SELECTION";
        CardReaderTransactionStatusProgress["INITIATING_APPLICATION"] = "INITIATING_APPLICATION";
        CardReaderTransactionStatusProgress["READING_APP_DATA"] = "READING_APP_DATA";
        CardReaderTransactionStatusProgress["OFFLINE_DATA_AUTHENTICATION"] = "OFFLINE_DATA_AUTHENTICATION";
        CardReaderTransactionStatusProgress["PROCESS_RESTRICTIONS"] = "PROCESS_RESTRICTIONS";
        CardReaderTransactionStatusProgress["CARDHOLDER_VERIFICATION"] = "CARDHOLDER_VERIFICATION";
        CardReaderTransactionStatusProgress["TERMINAL_RISK_MANAGEMENT"] = "TERMINAL_RISK_MANAGEMENT";
        CardReaderTransactionStatusProgress["TERMINAL_ACTION_ANALYSIS"] = "TERMINAL_ACTION_ANALYSIS";
        CardReaderTransactionStatusProgress["GENERATING_FIRST_APP_CRYPTOGRAM"] = "GENERATING_FIRST_APP_CRYPTOGRAM";
        CardReaderTransactionStatusProgress["CARD_ACTION_ANALYSIS"] = "CARD_ACTION_ANALYSIS";
        CardReaderTransactionStatusProgress["ONLINE_PROCESSING"] = "ONLINE_PROCESSING";
        CardReaderTransactionStatusProgress["WAITING_ONLINE_PROCESSING_RESPONSE"] = "WAITING_ONLINE_PROCESSING_RESPONSE";
        CardReaderTransactionStatusProgress["TRANSACTION_COMPLETION"] = "TRANSACTION_COMPLETION";
        CardReaderTransactionStatusProgress["TRANSACTION_ERROR"] = "TRANSACTION_ERROR";
        CardReaderTransactionStatusProgress["TRANSACTION_APPROVED"] = "TRANSACTION_APPROVED";
        CardReaderTransactionStatusProgress["TRANSACTION_DECLINED"] = "TRANSACTION_DECLINED";
        CardReaderTransactionStatusProgress["BUSY"] = "BUSY";
    })(CardReaderTransactionStatusProgress = KardReaderModule.CardReaderTransactionStatusProgress || (KardReaderModule.CardReaderTransactionStatusProgress = {}));
    var CardReaderDevice;
    (function (CardReaderDevice) {
        CardReaderDevice["EDYNAMO"] = "EDYNAMO";
        CardReaderDevice["DYNAWAVE"] = "DYNAWAVE";
        CardReaderDevice["MDYNAMO"] = "MDYNAMO";
        CardReaderDevice["INTELLIHEAD"] = "INTELLIHEAD";
    })(CardReaderDevice = KardReaderModule.CardReaderDevice || (KardReaderModule.CardReaderDevice = {}));
})(KardReaderModule = exports.KardReaderModule || (exports.KardReaderModule = {}));
var ReceiptPrinterModule;
(function (ReceiptPrinterModule) {
    var PaxPrinterErrorCode;
    (function (PaxPrinterErrorCode) {
        PaxPrinterErrorCode["OUT_OF_PAPER"] = "209904";
    })(PaxPrinterErrorCode = ReceiptPrinterModule.PaxPrinterErrorCode || (ReceiptPrinterModule.PaxPrinterErrorCode = {}));
})(ReceiptPrinterModule = exports.ReceiptPrinterModule || (exports.ReceiptPrinterModule = {}));
