// fivestars-pos-sdk/packages/types/lib/messenger/messenger.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PairedQueueFailedMessage = exports.TransportType = exports.Acknowledgment = void 0;
//import { ActionType } from 'typesafe-actions';
var ts_custom_error_1 = require("ts-custom-error");
//import * as messengerActions from './actions';
exports.Acknowledgment = 'ack';
var TransportType;
(function (TransportType) {
    TransportType["BLUETOOTH"] = "Bluetooth";
    TransportType["UDP"] = "UDP";
    TransportType["WEB_SOCKET"] = "WebSocket";
    TransportType["WEB_RTC"] = "WebRtc";
})(TransportType = exports.TransportType || (exports.TransportType = {}));
var PairedQueueFailedMessage = /** @class */ (function (_super) {
    __extends(PairedQueueFailedMessage, _super);
    function PairedQueueFailedMessage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PairedQueueFailedMessage;
}(ts_custom_error_1.CustomError));
exports.PairedQueueFailedMessage = PairedQueueFailedMessage;
