// fivestars-pos-sdk/packages/fivestars-pos-sdk-messenger/src/messenger/actions/transports/udp.ts
// import { createAction } from 'typesafe-actions';

// import { LOG_FILTER } from '@instore/utils/actions';

// export const reconnectRequest = createAction('UDP_RECONNECT_REQUEST');
export const reconnectRequest = () => ({ action: 'UDP_RECONNECT_REQUEST' })

// export const stopped = createAction('UDP_STOPPED');
export const stopped = () => ({ action: 'UDP_STOPPED' })

// export const receivedPairedDeviceBroadcast = createAction('UDP_RECEIVED_PAIRED_DEVICE_BROADCAST', resolve => () =>
//     resolve(LOG_FILTER)
// );
export const receivedPairedDeviceBroadcast = () => ({ action: 'UDP_RECEIVED_PAIRED_DEVICE_BROADCAST' })

// export const sendPing = createAction('UDP_SEND_PING', resolve => () => resolve(LOG_FILTER));
export const sendPing = () => ({ action: 'UDP_SEND_PING' })

// export const pongReceived = createAction('UDP_PONG_RECEIVED', resolve => () => resolve(LOG_FILTER));
export const pongReceived = () => ({ action: 'UDP_PONG_RECEIVED' })

// export const updateUdpPort = createAction('MESSENGER_UDPATE_UDP_PORT', resolve => (udpPort: number) =>
//     resolve(udpPort)
// );
export const updateUdpPort = (port: number) => ({ action: 'MESSENGER_UDPATE_UDP_PORT', data: port })
