// fivestars-pos-sdk/packages/fivestars-pos-sdk-messenger/src/messenger/actions/transports/webSocket.ts
// import { createAction } from 'typesafe-actions';

// import { WebSocketReconnectReason } from '@instore/middleware/messenger/transports/websocket/types';
// import { LOG_FILTER } from '@instore/utils/actions';
import {
    WebSocketReconnectReason
 } from '@fivestars-pos-sdk/types'

// export const errorEvent = createAction('WEB_SOCKET_ERROR_EVENT', resolve => (message: string) => resolve(message));
export const errorEvent = (message: string) => ({ action: 'WEB_SOCKET_ERROR_EVENT', data: message})

// export const reconnectRequest = createAction(
//     'WEB_SOCKET_RECONNECT_REQUEST',
//     resolve => (reason: WebSocketReconnectReason) => resolve({ reason })
// );
export const reconnectRequest = (reason: WebSocketReconnectReason) => ({ action: 'WEB_SOCKET_RECONNECT_REQUEST', data: reason})

// export const failedToReconnect = createAction('WEB_SOCKET_FAILED_TO_RECONNECT');
export const failedToReconnect = () => ({ action: 'WEB_SOCKET_FAILED_TO_RECONNECT' })

// export const connectEvent = createAction('WEB_SOCKET_CONNECT_EVENT');
export const connectEvent = () => ({ action: 'WEB_SOCKET_CONNECT_EVENT' })

// export const closeEvent = createAction('WEB_SOCKET_CLOSE_EVENT', resolve => (reason: string) => resolve(reason));
export const closeEvent = (reason: string) => ({ action: 'WEB_SOCKET_CLOSE_EVENT', data: reason})

// export const authenticationFailure = createAction('WEB_SOCKET_AUTHENTICATION_FAILURE');
export const authenticationFailure = () => ({ action: 'WEB_SOCKET_AUTHENTICATION_FAILURE' })

// export const authenticationSuccessful = createAction('WEB_SOCKET_AUTHENTICATION_SUCCESSFUL');
export const authenticationSuccessful = () => ({ action: 'WEB_SOCKET_AUTHENTICATION_SUCCESSFUL' })

// export const pongReceived = createAction('WEB_SOCKET_PONG_RECEIVED', resolve => () => resolve(LOG_FILTER));
export const pongReceived = () => ({ action: 'WEB_SOCKET_PONG_RECEIVED' })

// export const pongTimeout = createAction('WEB_SOCKET_PONG_TIMEOUT');
export const pongTimeout = () => ({ action: 'WEB_SOCKET_PONG_TIMEOUT' })

// export const subscriberAuthenticated = createAction('WEB_SOCKET_SUBSCRIBER_AUTHENTICATED');
export const subscriberAuthenticated = () => ({ action: 'WEB_SOCKET_SUBSCRIBER_AUTHENTICATED' })
